src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -202,21 +216,116 @@ public boolean isOpen()
       public void close() throws IOException
       {
         open = false;
+        useDelegateSmooshWriter = false;
         internalFiles.put(name, new Metadata(currOut.getFileNum(), startOffset, currOut.getCurrOffset()));
 
         if (bytesWritten != currOut.getCurrOffset() - startOffset) {
-          throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
-        }
-        if (bytesWritten != size) {
-          throw new IOException(
-              String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+            throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
+          }
+          if (bytesWritten != size) {
+            throw new IOException(
+                String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
+            );
+          }        
+        if(!mergeOnProgress && !useDelegateSmooshWriter)
```

Comment:
This logic is pretty obtuse.  For someone who is reading this code, it's not really clear why the merge is happening.  Likely, you should merge if there are any completed files, so why wouldn't it just always iterate through all completed files and merge them?